compiler/p/codegen/OMRInstruction.cpp

```diff
@@ -59,8 +59,15 @@ OMR::Power::Instruction::Instruction(TR::CodeGenerator *cg, TR::InstOpCode::Mnem
 void
 OMR::Power::Instruction::remove()
    {
-   self()->getPrev()->setNext(self()->getNext());
-   self()->getNext()->setPrev(self()->getPrev());
+   if (self()->getPrev())
+      self()->getPrev()->setNext(self()->getNext());
+   else
+      self()->cg()->setFirstInstruction(self()->getNext());
+
+   if (self()->getNext())
+      self()->getNext()->setPrev(self()->getPrev());
+   else
+      self()->cg()->setAppendInstruction(self()->getPrev());
    }
```

Comment:
This looks identical to the base class implementation [1]. Do you think we can remove the Power specific one?

[1] https://github.com/eclipse/omr/blob/e1ac81b8a6a4e3022cddf032558fca1b4cb66bd7/compiler/codegen/OMRInstruction.cpp#L216-L238

Reply:
Ah, at first I didn't notice that this was an override of something defined in the base class... It looks like a number of other codegens (ARM, AArch64, and RISC-V) all have their own implementations copied from the original Power implementation.

While the base class implementation does look _similar_ to this new implementation, it looks like it doesn't handle the case of removing the first instruction. In that case, it fails to call `cg->setFirstInstruction(...)` and will leave the instruction list in an inconsistent state.

IMO we should remove all codegen overrides of this method and replace the base class implementation with this new version. However, considering that this would touch other codegens and the changes I'm doing here are already very extensive, I'm hesitant to do that in this PR. Perhaps an issue should be opened so this can be handled after this PR merges?

Reply:
Sure, that sounds reasonable. I agree we should all move towards a common implementation and that can be done in a separate PR.

Reply:
Opened #5129 to track this.